### ISSUE TITLE
merge_artifacts loads objects into memory

### ISSUE THREAD

**Issue by peterwilton-ttam** [Netflix/metaflow#555]
I was surprised to see that `FlowSpec.merge_artifacts` loads all of the merged artifacts into memory after doing all of the consistency checks with hashes. Is there any reason it’s necessary to load all of the artifacts into memory? Would it instead be possible to copy over hashes and metadata from the input-branches' datastores into the join-step's datastore? That is, would it be possible to avoid `getattr` and `setattr` in `merge_artifacts`?

I have a flow with a join step that does nothing more than merge lots of artifacts of different names from upstream branches. I expected this to be a low-memory operation, but of course I'm finding I need to request enough memory to get all of these artifacts into memory. In my case this is a substantial request. And of course decompressing, depickling, repickling, and recompressing takes some time, too.

Here's a small patch of `merge_artifacts` that seems to accomplish this:

``` patch
-        for var, (inp, _) in to_merge.items():
-            setattr(self, var, getattr(inp, var))
+        for var, (inp, sha) in to_merge.items():
+            if self._datastore.object_exists(sha):
+                self._datastore.objects[var] = sha
+                self._datastore.info[var] = inp._datastore.info[var]
+            else:
+                setattr(self, var, getattr(inp, var))
```

where `LocalDataStore.object_exists` and `S3DataStore.object_exists` are defined in the expected way.

I'm having some difficulty getting all of the `dev-local` tests to pass on `master` (three tests are failing, and it seems it has to do with my local Python environments), but those tests that do pass on `master` also pass with these changes.

Is this something you've considered before? Am I missing something that would break with such changes?

Thanks!

**Comment by romain-intel** [comment]
Hi. This is an excellent suggestion (and don't know how I didn't see this). I think we may also be able to use the `passdown_partial` method in the datastore to do this. May I ask, however, why the check on `object_exists`? If it is present in one of the inputs, it *should* be present in the datastore (or am I missing something).

**Comment by peterwilton-ttam** [comment]
Great, thanks for the response.

> May I ask, however, why the check on `object_exists`?

I wasn't sure whether it was safe to assume that the input datastore is always configured like the join-step datastore. I suppose it is (it's not possible to have one step backed by a local datastore and the next by S3, say), in which case I agree it would be fine to just copy over the book-keeping without checking for existence.

**Comment by peterwilton-ttam** [comment]
I'd be happy to put in a PR from the [feature branch](https://github.com/peterwilton-ttam/metaflow/tree/merge-artifacts-without-loading) on my fork, if that would be helpful.

I wasn't familiar with `passdown_partial`, but looking at it it does look it does pretty much this.

**Comment by romain-intel** [comment]
Sure. That'd be great! And yes, passdown_partial is used when dealing with parameters which behaves in a similar way.

**Comment by romain-intel** [comment]
This issue was solved by #556.
